lib/src/front_end/ast_node_visitor.dart

```diff
@@ -1940,8 +1973,12 @@ class AstNodeVisitor extends ThrowingAstVisitor<Piece> with PieceFactory {
   /// If [commaAfter] is `true`, looks for a comma token after [node] and
   /// writes it to the piece as well.
   @override
-  Piece nodePiece(AstNode node, {bool commaAfter = false}) {
+  Piece nodePiece(AstNode node,
+      {bool commaAfter = false, NodeContext context = NodeContext.none}) {
+    var previousContext = _parentContext;
+    _parentContext = context;
     var result = node.accept(this)!;
+    _parentContext = previousContext;
```

Comment:
Will this replace `.parent` calls? Like in `visitConditionalExpression`? It would be nice to have one consistent way of finding out the parent context.

I also was wondering if we wanted each piece to have a parent context (or a parent) they could look at. It seems like parent context should exist on each Piece rather than in the visitor.

Reply:
> Will this replace `.parent` calls? Like in `visitConditionalExpression`? It would be nice to have one consistent way of finding out the parent context.

That's my aspiration, yes. I'll write a long TODO comment that effect.

> I also was wondering if we wanted each piece to have a parent context (or a parent) they could look at. It seems like parent context should exist on each Piece rather than in the visitor.

Maybe, I but I try to avoid storing state in Pieces unless we actually need it to persist all the way to line splitting. So far, I think we only really need the parent context during visiting and piece construction and it can be discarded afterwards, which is why I made it an ephemeral bit of state in the visitor.